mart/attack/adversary.py

```diff
@@ -110,13 +111,28 @@ def perturber(self) -> Perturber:
     def configure_optimizers(self):
         return self.optimizer(self.perturber)
 
+    def get_input_adv(self, *, input, target):
+        perturbation = self.perturber(input=input, target=target)
+        input_adv = self.composer(perturbation, input=input, target=target)
+        return input_adv
+
     def training_step(self, batch, batch_idx):
+        # TODO: We shouldn't need to copy because it is never changed?
         # copy batch since we modify it and it is used internally
-        batch = batch.copy()
+        # batch = batch.copy()
+
+        input = batch["input"]
+        target = batch["target"]
+        # What we need is a frozen model that returns (a dictionary of) logits, or losses.
+        model = batch["model"]
+
+        # Compose input_adv from input, then give to model for updated gain.
+        input_adv = self.get_input_adv(input=input, target=target)
+        # Target model expects input in the original format.
+        batch_adv = (input_adv, target)
 
-        # We need to evaluate the perturbation against the whole model, so call it normally to get a gain.
-        model = batch.pop("model")
-        outputs = model(**batch)
+        # A model that returns output dictionary.
+        outputs = model(batch_adv)
```

Comment:
Why are we not just (more less) calling `model.attack_step(batch, batch_idx)` here?

Reply:
Sure but instead of wrapping the model below, you can just detect if the model supports `attack_step` and use that here. Otherwise you default to `__call__`.

Reply:
It should be good to avoid `if` here if we decide to keep `model_transform` below?